gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/ReminderSettingDagProcLeaseArbiter.java

```diff
@@ -0,0 +1,172 @@
+ * that fail to acquire a lease will use the {@link DagActionReminderScheduler} to set a reminder for the flow action
+ * event to check back in on the previous lease owner's completion status.
+ */
+@Slf4j
+public class ReminderSettingDagProcLeaseArbiter implements MultiActiveLeaseArbiter {
+  private final MultiActiveLeaseArbiter decoratedLeaseArbiter;
+  private final DagActionReminderScheduler _dagActionReminderScheduler;
+  private final Config config;
+
+  @Inject
+  public ReminderSettingDagProcLeaseArbiter(Config config, MultiActiveLeaseArbiter leaseArbiter, DagActionReminderScheduler dagActionReminderScheduler) throws IOException {
+    this.decoratedLeaseArbiter = leaseArbiter;
+    this._dagActionReminderScheduler = dagActionReminderScheduler;
+    this.config = config;
+  }
+
+  /**
+   * This method is used by the multi-active scheduler and multi-active execution classes (DagTaskStream) to attempt a
+   * lease for a particular job event and return the status of the attempt.
+   * @param flowAction
+   * @param eventTimeMillis
+   * @param isReminderEvent
+   * @param skipFlowExecutionIdReplacement
+   * @return
+   */
+  @Override
+  public MultiActiveLeaseArbiter.LeaseAttemptStatus tryAcquireLease(DagActionStore.DagAction flowAction, long eventTimeMillis,
+      boolean isReminderEvent, boolean skipFlowExecutionIdReplacement) {
+    try {
+      MultiActiveLeaseArbiter.LeaseAttemptStatus leaseAttemptStatus = this.tryAcquireLease(flowAction, eventTimeMillis, isReminderEvent, skipFlowExecutionIdReplacement);
+      /* Schedule a reminder for the event unless the lease has been completed to safeguard against case lease owner
+      fails to complete lease
+      */
+      if (leaseAttemptStatus instanceof MultiActiveLeaseArbiter.LeaseObtainedStatus) {
+        scheduleReminderForEvent((LeaseObtainedStatus) leaseAttemptStatus);
+      } else if (leaseAttemptStatus instanceof  MultiActiveLeaseArbiter.LeasedToAnotherStatus) {
+        scheduleReminderForEvent((LeasedToAnotherStatus) leaseAttemptStatus);
+      }
+      return leaseAttemptStatus;
+    } catch (SchedulerException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public boolean recordLeaseSuccess(LeaseObtainedStatus status)
+      throws IOException {
+    return this.decoratedLeaseArbiter.recordLeaseSuccess(status);
+  }
+
+  protected void scheduleReminderForEvent(MultiActiveLeaseArbiter.LeasedToAnotherStatus leaseStatus)
+      throws SchedulerException {
+    _dagActionReminderScheduler.scheduleReminder(leaseStatus.getDagAction(), leaseStatus.getMinimumLingerDurationMillis());
+  }
+
+  protected void scheduleReminderForEvent(MultiActiveLeaseArbiter.LeaseObtainedStatus leaseStatus)
+      throws SchedulerException {
+    _dagActionReminderScheduler.scheduleReminder(leaseStatus.getDagAction(), leaseStatus.getMinimumLingerDurationMillis());
+  }
```

Comment:
I don't like the duplication... and I agree there's no perfect solution here (in the java type system).

but casting's imperfect on its own, so I'd probably bite the bullet and add *concrete impls* to the `LeaseAttemptStatus` that both return `null`:
```
public DagAction getDagAction() ...

public long getMinimumLingerDurationMillis() ...
```
thereafter the `@Data` or `@Getter`s on these two derived classes will override to a real value for each.

thereafter, the conditions above become merely:
```
if (!leaseAttemptStatus instanceof MultiActiveLeaseArbiter.NoLongerLeasing) {
  scheduleReminderForEvent(leaseAttemptStatus);
}
...
```